### PR TITLE
Provide Enum.chunk_by/4 and Stream.chunk_by/4

### PR DESCRIPTION
**Pull request by yagooar**
As requested by @josevalim in #5888, here is the initial changes I've made in order to implement `Stream.chunk_by/4`.

### CODE REVIEW

File: lib/elixir/lib/stream.ex [hunk 1]
```diff
@@ -185,16 +185,16 @@ defmodule Stream do
 
   ## Examples
 
-      iex> stream = Stream.chunk_by([1, 2, 2, 3, 4, 4, 6, 7, 7], &(rem(&1, 2) == 1))
```
Comment (josevalim) [inline lib/elixir/lib/stream.ex]: Examples are always indented with four spaces (it is markdown).

File: lib/elixir/lib/stream.ex [hunk 2]
```diff
@@ -205,6 +205,32 @@ defmodule Stream do
     next_with_acc(f1, :lists.reverse(buffer), h, nil, t)
   end
 
+  @doc """
+  Chunks the `enum` by buffering elements for which `fun` returns
```
Comment (josevalim) [inline lib/elixir/lib/stream.ex]: The first line of documentation should always be a short summary.

File: lib/elixir/lib/stream.ex [hunk 3]
```diff
@@ -205,6 +205,32 @@ defmodule Stream do
...
+  the same value and only emit them when `fun` returns a new value
+  or the `enum` finishes.
+
+  ## Examples
+
+  iex> start_func = fn i, acc -> if rem(i, 2) == 0, do: {Enum.reverse([i | acc]), []}, else: {[], [i | acc]} end
```
Comment (josevalim) [inline lib/elixir/lib/stream.ex]: Four spaces please.

File: lib/elixir/lib/stream/reducers.ex [hunk 4]
```diff
@@ -25,18 +25,35 @@ defmodule Stream.Reducers do
...
-        entry, acc(head, {buffer, value}, tail) ->
-          new_value = unquote(callback).(entry)
-          if new_value == value do
-            skip(acc(head, {[entry | buffer], value}, tail))
-          else
-            next_with_acc(unquote(fun), :lists.reverse(buffer), head, {[entry], new_value}, tail)
-          end
-        entry, acc(head, nil, tail) ->
-          skip(acc(head, {[entry], unquote(callback).(entry)}, tail))
+      fn(entry, acc(head, {buffer, value}, tail) = acc) ->
```
Comment (josevalim) [inline lib/elixir/lib/stream/reducers.ex]: Please don't change the code style. :) Keep it as before with `fn` on its own line and each clause indented two spaces.

**Comment by josevalim** [comment]
Hi @yagooar. I have provided some style feedback. I will jump into the code changes and give any feedback I may have.

File: lib/elixir/lib/stream/reducers.ex (hunk 4 above)
Comment (josevalim) [inline lib/elixir/lib/stream/reducers.ex]: Please don't change the code style. :) Keep it as before with `fn` on its own line and each clause indented two spaces.

**Comment by josevalim** [comment]
Hi @yagooar, I have pushed a working implementation of `Stream.chunk_by/4`. However, by doing that, I broke `Enum.chunk_by/2`. So now we need to add `Enum.chunk_by/4` and fix `Enum.chunk_by/2` accordingly. You can use the existing Enum.chunk_by/2 implementation for guidance as well as the new Stream ones. Please let me know if you have any questions.

**Comment by yagooar** [comment]
Thanks @josevalim for the review, will make sure to follow the styling rules more thoroughly :)
And thanks for the implementation, will work this afternoon on finishing the feature.

**Comment by yagooar** [comment]
Hey @josevalim, I have added a first implementation for `Enum.chunk_by/4`.

These are some questions that I still have:

What would be the expected output of the function, if I call:

```elixir
[3, 4, 5, 6, 7, 8, 9, 10] |> Enum.chunk_by([1, 2], chunk_fun, after_fun)
```

My bet is `[[1, 2], [3, 4], [5, 6], [7, 8], [9, 10]]`, but I'm not quite sure here since there is no example of this behaviour.

Next question is about adding `@type acc     :: any` to make `Enum` work. Could you explain why this works (I copied it over from `Stream`) and if it's the best solution?

Finally, I'd need some more guidance how to implement `chunk_by/2` based on `chunk_by/4`. The one in `Stream` is a bit cryptic to me right now, guess I'll need to play a bit more with it.

Thanks in advance for your patience, I really appreciate it :)

**Comment by josevalim** [comment]
> [3, 4, 5, 6, 7, 8, 9, 10] |> Enum.chunk_by([1, 2], chunk_fun, after_fun)

It should be the same result as in the Stream one, so you could try that one. :) I suspect it will be either the one you posted or `[[1, 2, 3, 4], [5, 6], ...]`.

**Comment by josevalim** [comment]
> Next question is about adding @type acc :: any to make Enum.chunk_by/4 work. Could you explain why this works (I copied it over from Stream) and if it's the best solution?

You don't need `@type acc :: any`. Just use `term` instead of `acc` in `chunk_by` typespec.

File: lib/elixir/lib/enum.ex [hunk 5]
```diff
@@ -376,6 +376,32 @@ defmodule Enum do
     end
   end
 
+  @type acc     :: any
+
+  @doc """
+  Splits enumerable on every element for which `fun` returns a new
```
Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Please copy the description from `Stream.chunk_by/4`.

File: lib/elixir/test/elixir/enum_test.exs [hunk 6]
```diff
@@ -66,6 +66,28 @@ defmodule EnumTest do
...
+      [] -> {:cont, []}
+      acc -> {:cont, Enum.reverse(acc), []}
+    end
+
+    assert [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] |> Enum.chunk_by([], chunk_fun, after_fun) ==
+      [[1, 2], [3, 4], [5, 6], [7, 8], [9, 10]]
+    assert 0..10 |> Enum.chunk_by([], chunk_fun, after_fun) ==
+      [[0], [1, 2], [3, 4], [5, 6], [7, 8], [9, 10]]
+    assert 0..11 |> Enum.chunk_by([], chunk_fun, after_fun) ==
+      [[0], [1, 2], [3, 4], [5, 6], [7, 8], [9, 10], [11]]
```
Comment (josevalim) [inline lib/elixir/test/elixir/enum_test.exs]: Please indent as in Stream.chunk_by/4 tests. :)

**Comment by josevalim** [comment]
Ping. :)

**Comment by yagooar** [comment]
Hey @josevalim, sorry for the delayed reply. Been a bit busy these days, will try to work on this today or tomorrow.

**Comment by josevalim** [comment]
Hi @yagooar, I will go ahead and finish this patch so we can continue moving forward with the v1.5 release. Thanks!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by yagooar** [comment]
Hey @josevalim, thanks for that. Next time I will try to make sure to have enough time to work on a PR :)

**Pull request merged**